src/mhd.f90

```diff
@@ -0,0 +1,907 @@
+  real(mytype),allocatable,dimension(:,:,:,:) :: Bm,magelf,Je
+  real(mytype),allocatable,dimension(:,:,:,:,:) :: dBm
+  real(mytype),allocatable,dimension(:,:,:) :: elcpot
+  !+------------+------------------------------------------------------+
+  !|          Bm| magnetic field                                       |
+  !|      magelf| Electromagnetic forced to apply to the momentum eq.  |
+  !|      elcpot| electric potential                                   |
+  !|          Je| electric field as the gradient of potential          |
+  !+------------+------------------------------------------------------+
+  !
+  contains
+  !
+  !+-------------------------------------------------------------------+
+  !| this subroutine is to initilise the MHD module.                   |
+  !+-------------------------------------------------------------------+
+  !| change record                                                     |
+  !| -------------                                                     |
+  !| 28-Oct-2022  | Created by J. Fang STFC Daresbury Laboratory       |
+  !+-------------------------------------------------------------------+
+  subroutine mhd_init
+    !
+    use param, only: re,ntime
+    !
+    ! stuart=hartmann**2/re
+    !
+    if(stuart<=1.d-15) then
+      stuart=hartmann**2/re
+    endif
+    if(hartmann<=1.d-15) then
+      hartmann=sqrt(stuart*re)
+    endif
+    !
+    if(nrank==0) then
+      !
+      print*,'** MHD Module activated'
+      print*,'**    MHD equation: ',mhd_equation
+      print*,'** Hartmann number: ',hartmann
+      print*,'**   Stuart number: ',stuart
+      print*,'**              Re: ',re
+      print*,'**     Magnetic Re: ',rem
+      !
+    endif
+    !
+    allocate( Bm(xsize(1),xsize(2),xsize(3),1:3),                  &
+              magelf(xsize(1),xsize(2),xsize(3),1:3),              &
+              Je(xsize(1),xsize(2),xsize(3),1:3),                  &
+              elcpot(xsize(1),xsize(2),xsize(3)),                  &
+              dBm(xsize(1),xsize(2),xsize(3),1:3,1:ntime) )
+    !
+    allocate(Bmean(xsize(1),xsize(2),xsize(3),1:3))
+    !
+    if(nrank==0) print*,'** MHD fields allocated'
+    !
+  end subroutine mhd_init
+  !+-------------------------------------------------------------------+
+  !| The end of the subroutine mhd_init.                               |
+  !+-------------------------------------------------------------------+
+  !
+  subroutine int_time_magnet
+    !
+    USE param
+    USE variables
+    USE decomp_2d_mpi, only : decomp_2d_abort
+    use ydiff_implicit
+
+    implicit none
+    !
+    integer :: k
+
+    if( iimplicit == 1 ) then
+       call decomp_2d_abort(iimplicit, "MHD is not compatible with implicit diffusion")
+    else
+       if(itimescheme.eq.3) then
+           !>>> Adams-Bashforth third order (AB3)
+
+           ! Do first time step with Euler
+           if(itime.eq.1.and.irestart.eq.0) then
```

Comment:
Major issue : I think MHD can not perform a restart because the field dBm is not written

Reply:
@rfj82982 I will fix this